File: go/sorting/shell_sort.go
package sorting

type ShellSort struct{}

func (s ShellSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)

	gap := 1
	for gap < length/9 {
		gap = gap*3 + 1
	}

	for gap > 0 {
		for i := gap; i < length; i++ {
			j := i
			for j >= gap && Compare(arr[j-gap], arr[j], sortType) {
				arr[j-gap], arr[j] = arr[j], arr[j-gap]
				j--
			}
		}
		gap = gap / 3
	}

	return arr
}

File: scala/src/main/scala/sorting/ShellSort.scala
package sorting

class ShellSort() extends Sort {
  def sort(arr: Array[Int], sortType: SortType): Array[Int] = {
    val length: Int = arr.length

    var gap = 1
    while (gap < length / 9) {
      gap = gap * 3 + 1
    }
    while (gap > 0) {
      for (i <- Range(gap, length, 1)) {
        var j = i
        while (j >= 1 && compare(arr(j - gap), arr(j), sortType)) {
          swap(arr, j, j - gap)
          j -= 1
        }
      }
      gap = gap / 3
    }
    arr
  }
}
